README.md
# chess-analysis
![](datasets/elo/elo_rapid.png)
This is an ongoing project, inspired from Lichess' analysis tools.

At the time of creation, Chess.com did not have an analysis feature, so I decided to implement some of the functions.

# features

## database 
Save all games in a database for later retrieval.
## engine analysis
All games are analyzed using Stockfish 14.
## heuristics
Custom heuristics, such as average centipawn loss, opportunistic/luckiness, and more are available for use.

# usage stuff
## database
### updating
- ```update_db.py``` gets games from the current month.
- To get from games further back in time, use ```get_games()``` in ```core.py``` manually.

## engine
### installation
- run `brew install stockfish`
### finding after installation
- make sure you are in CHESS_ANALYSIS, and run:
`cd /usr/local/opt/`

# structure
![](chess_analysis_map.png)

## core.py
The fundamental building block of the program.
- ```global``` ```string``` ```user```: username in PGNs
- ```global``` ```games_list```: an array containing all games (see ```python-chess```)

### DatabaseTool()
#### attributes
- conn: ```Connection``` a connection to the database
- cur: ```Cursor``` a cursor on the database
- full_collection: ```DataFrame``` all of the games
#### functions

```get()```: gets games from database
- ```filters```: ```dict``` all filters to apply
  - default ```None```
- ```db```: ```string``` table name to search
  - default ```"games"```

```get_by_column_name()```: gets games by column name
  - ```col_name```: ```string``` field name
  - ```table```: ```string``` table name

changelog.txt
DATE CREATED:
8/16/2021

GOAL:
to implement the features of LiChess' Chess Insight tab locally on my chess games

AIMS:
-to practice efficient implementation of data analysis tools
-to practice the things I'm currently learning about in computer science

TIMEFRAME
8/16/2021 2:47AM created folder datasets
8/16/2021 3:00AM downloaded all live chess games from chess.com account roudiere
	--> used opening tree to collect all files then export as pgn
	--> result was two PGNs, roudiere-black.pgn and roudiere-white.pgn
	--> need to combine these into one for a condensed data set (easier querying)
8/16/2021 3:20AM merged roudiere-black.pgn and roudiere-white.pgn into one PGN, roudiere.pgn using merge_black_white_files.py
8/16/2021 3:26AM merged roudiere-black.pgn and roudiere-white.pgn into one PGN, roudiere.pgn using zsh
    --> zsh command: "cat * > roudiere.pgn"
    --> learned using this reference: https://unix.stackexchange.com/questions/3770/how-to-merge-all-text-files-in-a-directory-into-one
    --> demonstrates the importance of shell -- gives access to a whole host of tools that are not accessible by GUI because of cluttering
8/16/2021 3:30AM moved merge_black_white_files.py to deprecated (for educational purposes)
    --> demonstrates file pointers, basic python
8/16/21 3:57AM began research on python-chess
    --> https://readthedocs.org/projects/python-chess/downloads/pdf/latest/
    --> data structures already provided, eval writing already provided (page 43)
    --> skimming: useful for providing queries (page 47)
    --> can communicate with UCI engines (e.g. Stockfish)
8/17/21 1:40AM installed python-chess as dependency
8/17/21 2:02AM created chess_analysis.py
8/17/21 2:12AM created total_games() in chess_analysis.py
8/17/21 2:16AM created color_games() in chess_analysis.py
8/17/21 2:25AM renamed total_games() to build_game_list()
8/17/21 2:25AM created get_total_games()
8/17/21 2:28AM created determine_time_control_numbers()
8/17/21 2:46AM installed stockfish 14 through homebrew
    --> brew install stockfish
8/17/21 3:03AM created iterate_through_moves()
    --> calculates the score for each move using Stockfish
        --> prints the score from player's POV
    --> needed to find the folder that contained stockfish: "/usr/local/opt/stockfish/bin/stockfish"
8/17/21 4:02PM began creating evaluations of all positions in database using python-chess and stockfish
8/17/21 9:35PM finished evaluations -- seems accurate
8/20/21 3:12AM created get_total_games_by_filter()
    --> ended up finding two elements with no moves played -- removed from dataset
    --> cleaned up evaluations.txt as well -- found two blank spots in file as well as two with one evaluation
        --> guessing those are 1 move wins when I played white?
8/20/21 3:37AM installed pandas, seaborn
8/20/21 3:58AM basic data analysis
    --> used get_total_games_by_filter() to find most played openings
    --> used evaluation data from rligloo vs roudiere to create an advantage graph
8/20/21 3:00PM basic data analysis
    --> moved code from advantage graph to plot_evaluation_chart()
    --> created basic_stats_from_eval()
8/21/21 9:03AM data analysis
    --> first attempt at answering the question, "how well do I play from each opening?"
        --> calculated average (cp_loss, inaccuracies, mistakes, blunders) for every opening played
        --> moved most played openings code to get_most_played_openings()
        --> moved code to answer question to calc_all_openings_basic_acc()
8/21/21 6:30PM created and filled database to hold all games
8/21/21 7:47PM created get_all_forced_mates()
    --> creates a text file with the FENs of all of the positions with forced mates
8/21/21 8:02PM improved get_all_forced_mates()
    --> takes optional input, till_mate, which only looks for positions with mate in N
8/23/21 2:52AM answered the question, "What percentage of my games make it to the endgame?"
    --> approximately 55%
8/23/21 3:09AM answered the question, "What percentage of my games end in the opening?" (less than 12 moves)
    --> approximately 8%
8/23/21 3:29AM completed analysis of all stages of the game
    opening ------ 7.67% end rate, 56.04% win rate
    middlegame -- 37.40% end rate, 52.25% win rate
    endgame ----- 54.93% end rate, 51.53% win rate
    --> so it seems my opening is my strongest point, and I need to focus on endgames as most of my games finish there
    --> another insight: I am a sharp player who aims for quick wins. The longer the game goes on, the better of a chance
        my opponent has of winning (assuming equal strength)
8/23/21 6:09PM created two measures of games from opening analysis:
    --> comfortableness: games / (avg_cp_loss)**2
        --> the more games I play, the more comfortable I am with the opening
        --> the worse the cp_loss, the more likely I am to be unfamiliar with the positions
    --> sharpness: win_percent / avg_cp_loss
        --> measure of how efficient I am in winning
8/24/21 4:18AM changed opening analysis to save opening ECO code instead of URL
    --> should result in better grouping, since chess.com breaks openings up into more lines than I require
8/24/21 4:19AM downloaded list of ECO codes from chessgames.com
    --> URL: https://www.chessgames.com/chessecohelp.html
8/24/21 4:22AM installed bs4 (BeautifulSoup)
8/24/21 4:33AM created eco_codes.csv
    --> CSV file which contains all ECO codes and their names
    --> https://stackoverflow.com/questions/18966368/python-beautifulsoup-scrape-tables
8/24/21 5:45AM replaced names of ECO code with ECO code -- name (line)
    --> got hung up on bug for an hour: turns out the code I pasted in arbitrarily threw away the first two elements
8/24/21 6:28AM cleaned up data analysis code, created folders for files
    --> core: stuff such as the original PGN files, the merged file, the evaluation file
    --> mates: data gathered/analysis on mates
    --> openings: data gathered/analaysis on openings

basic_stats.py
import core
import numpy as np

def count_total_games(f=None):
    print("total games:", len(core.dbt.get(f)))

def count_moves(g):
    j = 0
    board = g.board()
    for move in g.mainline_moves():
        board.push(move)
        j += 1

    return np.ceil(j / 2), board

def basic_stats_from_eval(eval_lst, color):
    stats = {"cp_loss": [],
             "avg_cp_loss": 0,
             "inaccuracies": 0,
             "mistakes": 0,
             "blunders": 0}

    if color.lower() == 'white':
        stats["cp_loss"] = [abs(eval_lst[i + 1] - eval_lst[i]) if eval_lst[i + 1] < eval_lst[i] else 0
                            for i in range(1, len(eval_lst) - 1, 2)]
    elif color.lower() == 'black':
        stats["cp_loss"] = [abs(eval_lst[i + 1] - eval_lst[i]) if eval_lst[i + 1] > eval_lst[i] else 0
                            for i in range(0, len(eval_lst) - 1, 2)]

    stats["avg_cp_loss"] = np.nan_to_num(np.average(stats["cp_loss"]) / 100)

    # calculate inaccuracies, mistakes, blunders
    for i, cpl in enumerate(stats["cp_loss"]):
        # don't count inaccuracies, mistakes, blunders if a move is made that directly increases the advantage

        # move was more positive than the one before it
        if color == 'white' and eval_lst[2 * (i + 1)] > eval_lst[2 * (i + 1) - 1]:
            continue
        # move was more negative than the one before it
        elif color == 'black' and eval_lst[2 * (i + 1) - 1] < eval_lst[2 * (i + 1) - 2]:
            continue

        if cpl < 100:
            continue
        elif 100 <= cpl < 150:
            stats["inaccuracies"] += 1
        elif 150 <= cpl < 400:
            stats["mistakes"] += 1
        elif cpl >= 400:
            stats["blunders"] += 1

    return stats

def clean_eval_list(n):
    """
    Turn the nth evaluation list in chess_games.db from a string into a sequence of numbers
    especially fitted to encode smoothness with respect to calculating centipawn loss.
        e.g. "+12,+25,+47,+21,-65,-45,-300,-3500,-#3,-250,-249,+300,+#1,+#0" gets transformed into
             [12,25,47,21,-65,-45,-300,-3500,-3500,-250,-249,+300,+1800,+1800]

    Runtime: O(n)

    :param n: game number to choose
    :return: list of evaluations (ints)
    """
    # mark mates
    cp_limits = 1800

    eval_lst = core.evals[n - 1].split(",")

    eval_lst = [_ for _ in eval_lst if _ != '']
    eval_lst = [_ if "-#" not in _ else "black_mate" for _ in eval_lst]
    eval_lst = [_ if "#" not in _ else "white_mate" for _ in eval_lst]

    # calculate biggest advantages for either side
    biggest_black_adv = min([0 if _ in ["black_mate", "white_mate"] else int(_) for _ in eval_lst])
    biggest_white_adv = max([0 if _ in ["black_mate", "white_mate"] else int(_) for _ in eval_lst])

    # set mate to either max of position or 18 pawns if position is worse than 18 pawns (2 queens)
    eval_lst = [_ if _ != "black_mate" else str(min(-cp_limits, biggest_black_adv)) for _ in eval_lst]
    eval_lst = [_ if _ != "white_mate" else str(max(cp_limits, biggest_white_adv)) for _ in eval_lst]

    # convert to integers
    eval_lst = [int(_[1:]) if "+" in _ else int(_) for _ in eval_lst]

    return eval_lst

def get_test_game(x, color):
    player = "White" if color.lower() == 'white' else "Black"
    # get test game's eval
    return core.dbt.full_collection.iloc[x - 1][player], basic_stats_from_eval(clean_eval_list(x), color)

def get_stats_by_link(link):
    df = core.dbt.full_collection
    for index, row in df.iterrows():
        if df.iloc[index]["Link"] == link:
            white, w_stats = get_test_game(index + 1, 'white')
            black, b_stats = get_test_game(index + 1, 'black')

            return (white, w_stats), (black, b_stats)

    return -1